keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,813 @@
+These restrictions are derived from [RFC-1123](https://datatracker.ietf.org/doc/html/rfc1123). However,
+DNS name segments may contain more characters than are allowed in host names (see [RFC-2181](https://datatracker.ietf.org/doc/html/rfc2181#section-11)).
+In practice, DNS names sometimes contain characters, such as underscores, which would not be allowed by RFC-1123.
+
+This KEP proposes relaxing the checks on DNS search strings, specifically, without changing the restrictions on Kubernetes object
+names or other uses of DNS names within Kubernetes. Allowing these values in the `searches` field of `dnsConfig` allows pods to
+resolve short names properly in cases where the search string contains characters that are not currently allowed.
+
+## Motivation
+
+For workloads that resolve short DNS names where the full DNS name includes disallowed characters (like underscores),
+it’s not possible to configure search strings using dnsConfig. For example, if a pod needs to resolve abc123.abc_d.example.com
+using the short name of abc123, we would like to be able to add abc_d.example.com to the searches in the dnsConfig.
+
+Here’s an example configuration which would support this case:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: dns-example
+spec:
+  containers:
+    - name: test
+      image: nginx
+  dnsPolicy: "None"
+  dnsConfig:
+    nameservers:
+      - 1.2.3.4
+    searches:
+      - abc_d.example.com
+```
+
+However, this returns an error:
+
+```
+The Pod "dns-example" is invalid: spec.dnsConfig.searches[0]: Invalid value: "abc_d.example.com": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
+```
+
+While a case could be made that these names are incorrect as host names based on RFC-1123, DNS servers generally allow such names
+and once they are in use DNS names can be complicated to change especially when they're used by legacy systems and not purely
+within Kubernetes. The DNS searches in the Kubernetes pods spec are not defining the host name, so even under the stricter interpretation
+of RFC-1123, relaxing these checks does not allow users of Kubernetes to introduce an invalid host name where one did not
+already exist outside of Kubernetes.
+
+### Goals
+
+- Support workloads that need to resolve DNS short names where the full DNS name includes characters disallowed by RFC-1123.
+
+## Proposal
+
+Introduce a RelaxedDNSSearchValidation feature gate which is disabled by default. When the feature gate is enabled,
+DNS search strings will be validated according to RFC-2181:
+- The length of any one label is limited to between 1 and 63 octets
+- A full domain name is limited to 255 octets (including the separators)
```

Comment:
Can you please list explicitly what characters are allowed or disallowed?  i.e. a label may not container a NUL (`'\0'`) or a dot (`'.'`).  Is there anything else?  You can cite the specific RFC section and summarize.

Reply:
Thanks.  This sounds right, but I'm always wary of going too big too fast. Maybe we should restrict it to ASCII printable characters, minus whitespace, minus period?  If there's need to have more than that, I'd love an opportunity to understand it before implementing.

Reply:
That would definitely cover our case.

Resolvers appear to pass bytes through as they appear in the file.

For example, from golangs resolver:

```golang
func getFields(s string) []string { return splitAtBytes(s, " \r\t\n") }
```
[Source](https://github.com/golang/go/blob/4afb155cdf5cd6e8b133174d3369c4db4ea23c77/src/net/parse.go#L123)

```golang
name := ensureRooted(f[i])
if name == "." {
   continue
}
conf.search = append(conf.search, name)
```
[Source](https://github.com/golang/go/blob/4afb155cdf5cd6e8b133174d3369c4db4ea23c77/src/net/dnsconfig_unix.go#L68-L72)

Checked glibc and it works similarly.

That said, i think it's unlikely anyone is relying on unprintable characters in the resolv.conf text file and there may be some issue lurking with allowing those. I'm fine with ASCII printable, minus whitespace, minus period.